feat: add web flag for codespace list & create subcommand

doaortu opened the pull request:
This PR could fix #7261 

### Changes were made:
<br>

#### list subcommand

- adding web flag for opening list of codespace in browser with url: `https://github.com/codespaces`.
- `--web` flag can only be used with `--repo` flag, because in `https://github.com/codespaces` there only GET param for filtering with repository_id. (no params to filter with user or organizations). It will throws error if used with `--user` or `org` flag
- when user only specify `--web` flag without additional flags, it will open browser with default url: `https://github.com/codespaces`
- when user specify `--web` flag with `--repo` flag, it will first fetch repo_id from the api, and then open browser with url+repo_id param: `https://github.com/codespaces?repository_id={id}`
- add test case this newly added `--web` flag with above scenario
- modify the `t.Run` clause to add the `&browser.Stub{}` if `tt.opts.useWeb` is `true` (on `pkg/cmd/codespace/list_test.go`)

<br>

#### create subcommand

- adding web flag for creating codespace through browser with url: `https://github.com/codespaces`
- `--web` flag cannot be used with `--display-name`, `--idle-timeout`, or '--retention-period' flag because there's no option to set those in browser/Web UI. when used with those flags, it will throws error
- when user only specify `--web` flag without additional flags, it will open browser with default url: `https://github.com/codespaces/new`
- when user specify `--web` flag with `--repo` flag, it will first fetch repo_id from the api, and then open browser with url+repo_id param: `https://github.com/codespaces/new?repo={id}&ref={defaultBranch}&machine={firstMachineInList}&location={emptyString}`. 
- when no machine specified, it will automatically select the first machine instead of displaying machine chooser in terminal. (because already got machine chooser in web UI)
- when no branch specified, it will automatically select the default branch for the repository
- when no location specified, it will yield empty string to the location URL param, because the Web UI will automatically choose default location if none specified.
- when additional flags specified (repo, branch, machine, location) , it will plug the value to the respective params: `https://github.com/codespaces/new?repo={id}&ref={specifiedBranch}&machine={specifiedMachine}&location={specifiedLocation}`
- add test case this newly added `--web` flag with above scenario
- modify the `t.Run` clause to add the `&browser.Stub{}` if `tt.opts.useWeb` is `true` (on `pkg/cmd/codespace/create_test.go`)

doaortu committed:
feat: add web flag to cs list subcommand

web flag only works with repo flag, because,
currently there only param for listing with repo_id

doaortu committed:
feat: add web flag to cs crate subcommand

web flag used for creating codespace through web UI instead of terminal.
web flag cannot be used with display-name, idle-timeout,
or retention retention-period
because there's no option for that in the Web UI

doaortu committed:
Merge branch 'cli:trunk' into feat/web-flag-for-cs-ls-create

jkeech commented:
Thanks for opening this PR! Tagging @cmbrose to help review.

doaortu committed:
Merge branch 'trunk' into feat/web-flag-for-cs-ls-create

doaortu committed:
Merge branch 'cli:trunk' into feat/web-flag-for-cs-ls-create

doaortu commented:
Thank you. Please let me know if there's anything I can help

cmbrose commented on pkg/cmd/codespace/create.go:
```diff
@@ -271,6 +281,10 @@ func (a *App) Create(ctx context.Context, opts createOptions) error {
 		DisplayName:            opts.displayName,
 	}
```
It looks like if we don't set `machine` in the query params (just use `&machine=`) then the browser selects the correct default. In that case, could we skip this machine check?

doaortu replied:
Hi, thanks for detailed review.

Ah yeah, you're right. Will fix it real quick with other issues you mentioned in this PR.

doaortu replied:
fixed in [1f41686](https://github.com/cli/cli/pull/7288/commits/1f41686e778b0c822e8804f2671b04c77faf85ec) .
please let me know if there's something I need to change.

cmbrose commented on pkg/cmd/codespace/create.go:
```diff
@@ -118,7 +120,15 @@ func (a *App) Create(ctx context.Context, opts createOptions) error {
 	}
 
-	promptForRepoAndBranch := userInputs.Repository == ""
+	if opts.useWeb && (opts.displayName != "" || opts.idleTimeout != 0 || opts.retentionPeriod.Duration != nil) {
+		return errors.New("using --web with --display-name, --idle-timeout, or --retention-period is not yet supported")
+	}
```
Instead of doing the mutual exclusion check inside the main command, the usual pattern in the cli is to either use the `cobra` library's helper:

```go
createCmd.Flags().BoolVarP(&opts.useWeb, "web", "w", false, "create codespace from browser, cannot be used with --display-name, --idle-timeout, or --retention-period")

if err := createCmd.MarkFlagsMutuallyExclusive("web", "display-name"); err != nil {
   return err;
}
// repeat for other options
```

or use the custom `cmdutil.MutuallyExclusive` helper

```go
createCmd := &cobra.Command{
		// ...
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := cmdutil.MutuallyExclusive(
				"using --web with --display-name, --idle-timeout, or --retention-period is not yet supported",
				opts.useWeb,
				opts.displayName != "" || opts.idleTimeout != 0 || opts.retentionPeriod.Duration != nil,
			); err != nil {
				return err
			}

			return app.Create(cmd.Context(), opts)
		},
	}
```

I think in this case the resulting error is nicer using `cmdutil` as it will print the error message as well as show the `--help` text. Either way, it's better to put the mutual exclusion rules close to the flag definitions so that they're easier to see immediately

doaortu replied:
fixed in [c7db18a](https://github.com/cli/cli/pull/7288/commits/c7db18a3858bc60dfa33a9fd3ddc86e93afe369c) .
took me awhile to figure out how to test this. thankfully I figure it out by imitating `pkg/cmd/repo/create/create_test.go` (it also use `cmdutil.MutuallyExculsive` fn in `runE` clause)

please let me know if I need to change it further.

doaortu replied:
I also refactor the mutual exclusive flag check logic in `list.go` , I applied same change and add same test. 
It's on this [0798456](https://github.com/cli/cli/pull/7288/commits/079845623f0e7b52f3357992a102140b190c6deb) commit.

Please let me know what you think.

cmbrose commented on pkg/cmd/codespace/create.go:
```diff
@@ -118,7 +120,15 @@ func (a *App) Create(ctx context.Context, opts createOptions) error {
 
-	promptForRepoAndBranch := userInputs.Repository == ""
+	if opts.useWeb && (opts.displayName != "" || opts.idleTimeout != 0 || opts.retentionPeriod.Duration != nil) {
+		return errors.New("using --web with --display-name, --idle-timeout, or --retention-period is not yet supported")
```
Nit: I don't think we want to say `yet` as it sounds like we plan to do it later, but that's not really the case right now. 

```suggestion
                return errors.New("using --web with --display-name, --idle-timeout, or --retention-period is not supported")
```

cmbrose commented on pkg/cmd/codespace/list.go:
```diff
@@ -53,10 +54,20 @@ func newListCmd(app *App) *cobra.Command {
 	listCmd.Flags().StringVarP(&opts.userName, "user", "u", "", "The `username` to list codespaces for (used with --org)")
 	cmdutil.AddJSONFlags(listCmd, &exporter, api.CodespaceFields)
 
+	listCmd.Flags().BoolVarP(&opts.useWeb, "web", "w", false, "List codespaces in the web browser")
```
Can this text point out the exclusivity with `--user` and `--org`?

doaortu replied:
fixed in [c4228f1](https://github.com/cli/cli/pull/7288/commits/c4228f11cb2f328dd4afe577c0ce1167e0b30dbe) .
I change it to : `"List codespaces in the web browser, can only be used with --repo flag"`
This commit also clarify other issue like:
- removing 'yet' in error messages

cmbrose commented on pkg/cmd/codespace/list.go:
```diff
@@ -53,10 +54,20 @@ func newListCmd(app *App) *cobra.Command {
 }
 
 func (a *App) List(ctx context.Context, opts *listOptions, exporter cmdutil.Exporter) error {
+	if opts.useWeb && (opts.orgName != "" || opts.userName != "") {
+		return cmdutil.FlagErrorf("using `--web` with `--org` or `--user` is not yet supported, please use with `--repo` instead")
+	}
```
Similar comment here as in `create` about `yet` as well as using the helpers for the mutual exclusion check

cmbrose commented on pkg/cmd/codespace/create.go:
```diff
@@ -271,6 +281,10 @@ func (a *App) Create(ctx context.Context, opts createOptions) error {
 	}
 
+	if opts.useWeb {
+		return a.browser.Browse(fmt.Sprintf("https://github.com/codespaces/new?repo=%d&ref=%s&machine=%s&location=%s", createParams.RepositoryID, createParams.Branch, createParams.Machine, createParams.Location))
```
In browser, if you select `New with options` for the `Code` menu of a repo, we add  `hide_repo_select=true` to the query params so that the repo picker is hidden. Since the user has already selected the repo in this case, should we use that flag that here also?

doaortu replied:
Ah, I don't know there's `hide_repo_select=true` param before. Thank you for pointing it out.

But is it better to keep showing the repo dropdown? Just in case the user wants to change the repo midway. 
How do you think?

cmbrose replied:
I could see it either way, I would think that since the user already specified the repo when they ran the command they would have picked the one they wanted, but maybe that's not the case always. It's a minor thing, so I'm fine to not use the flag, was just letting you know it was an option.

doaortu replied:
Okay, I think it would be better to leave the repo select visible to maintain flexibility.

But still, if you think it's better to hide it, please let me know and I'll make it happen.

cmbrose reviewed (commented):
Thanks for working this up and for updating the tests! It looks solid, just a few thoughts on extra things we might want to include here

doaortu committed:
Merge branch 'cli:trunk' into feat/web-flag-for-cs-ls-create

doaortu committed:
refactor: extract mutual excusive logic to PreRunE

- changed web flag mutual exclusive logic, using cmdutil
- extract that logic to PreRunE clause in createCmd
- move web flag up to make it close to PreRunE clause (for clarity)
- add new param to newCreateCmd fn to facilitate test logic
- apply new newCreateCmd fn to root.go

doaortu committed:
fix: clarify flag desc and error message

- remove 'yet' from error messages that can cause misunderstanding
- clarify list web flag can only be used with repo flag

doaortu committed:
fix: skip machine check when we use web flag ...
(..and no machine flag provided)

+ add test for this new case
+ adjust related test cases for this new change.

doaortu committed:
refactor: move flag check logic to PreRunE

why: err on PreRunE or RunE will also print help if error happened

+ move web, repo, org, user mutual exclusive logic to PreRunE clause
+ move repo, org, user mutual exclusive logic to PreRunE
+ move limit check flag to PreRunE
+ modify newListCmd fn to facilitate test logic
+ apply new newListCmd to root.go
+ add test cases to check PreRunE clause
- remove mutual exclusive test cases from Test_AppList

cmbrose commented on pkg/cmd/codespace/create_test.go:
```diff
@@ -1,17 +1,83 @@
+				return
+			}
+
+			assert.Equal(t, tt.wantsOpts.branch, opts.branch)
+			assert.Equal(t, tt.wantsOpts.displayName, opts.displayName)
+			assert.Equal(t, tt.wantsOpts.idleTimeout, opts.idleTimeout)
+			assert.Equal(t, tt.wantsOpts.machine, opts.machine)
+			assert.Equal(t, tt.wantsOpts.repo, opts.repo)
+			assert.Equal(t, tt.wantsOpts.retentionPeriod, opts.retentionPeriod)
+			assert.Equal(t, tt.wantsOpts.useWeb, opts.useWeb)
```
I'm not sure if we really want this validation, I think that the error check above is enough to confirm the expected mutual exclusion rules.

The opts are an implementation detail though and we may refactor their structure over time which would cause these tests to break.

If we think that this validation is useful, we could instead use a mock `App` and override the `Create` function to capture the opts rather than needing to add the new `runF` parameter

doaortu replied:
Okay, I will remove these equal checks after this. Please bear with me.

doaortu replied:
fixed in [233d707](https://github.com/cli/cli/pull/7288/commits/233d707e640d3ee30924f16a86bef29e5102c106)

cmbrose commented on pkg/cmd/codespace/create.go:
```diff
@@ -69,20 +69,30 @@ type createOptions struct {
 		Short: "Create a codespace",
 		Args:  noArgsConstraint,
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			return cmdutil.MutuallyExclusive("using --web with --display-name, --idle-timeout, or --retention-period is not supported", opts.useWeb, opts.displayName != "", opts.idleTimeout != 0, opts.retentionPeriod.Duration != nil)
```
This doesn't work like you intend - `cmdUtil.MutuallyExclusive` just counts the number of `true` parameters and fails if there are more than 1. So if you don't set `--web`, but do set 2 of the other parameters it will fail: 

![image](https://user-images.githubusercontent.com/5447118/231508288-af910ac6-dfd7-48bd-a6bb-f4907357912f.png)

That's why I suggested `||`ing the additional parameters rather than supplying each individually.

doaortu replied:
Ah I'm sorry, I missed that. I also meant to do the `||` operation.
Will fix it after this

doaortu replied:
fixed in [99e2de2](https://github.com/cli/cli/pull/7288/commits/99e2de2c425905ef7fbe56db8a816a6d81ffd841)

cmbrose commented on pkg/cmd/codespace/create.go:
```diff
@@ -470,6 +495,11 @@ func getMachineName(ctx context.Context, apiClient apiClient, repoID int, machin
 		return machines[0].Name, nil
 	}
 
+	if useWeb {
```
In the `useWeb` case we now would only be in this method just to verify that the machine type exists - in that case we would have already returned in the above block because we either matched the machine or we didn't. So I don't think we need this `if` or the `useWeb` parameter anymore

doaortu replied:
Okay will remove the `useWeb` check and `useWeb` param from the `getMachineName` function

doaortu replied:
fixed in [d447e7a](https://github.com/cli/cli/pull/7288/commits/d447e7af91df5be4d6da2f80023784fccd5a13d8)

cmbrose commented on pkg/cmd/codespace/list.go:
```diff
@@ -34,10 +35,23 @@ func newListCmd(app *App) *cobra.Command {
+			if opts.repo != "" {
+				return cmdutil.MutuallyExclusive("using `--org` or `--user` with `--repo` is not allowed", opts.repo != "", opts.orgName != "" || opts.userName != "")
+			}
+			if opts.useWeb {
```
I don't think this `if` is needed - it's handled by the `MutuallyExclusive` check

doaortu replied:
fixed in [28bbd99](https://github.com/cli/cli/pull/7288/commits/28bbd99bd8948190d8b2fd3532ed014aa0207052)

cmbrose commented on pkg/cmd/codespace/list.go:
```diff
@@ -34,10 +35,23 @@ func newListCmd(app *App) *cobra.Command {
-		RunE: func(cmd *cobra.Command, args []string) error {
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			// if repo is provided, we don't accept orgName or userName
+			if opts.repo != "" {
```
I don't think this if is needed - it's handled by the MutuallyExclusive check

cmbrose commented on pkg/cmd/codespace/list.go:
```diff
@@ -53,13 +67,14 @@ func newListCmd(app *App) *cobra.Command {
 	listCmd.Flags().StringVarP(&opts.userName, "user", "u", "", "The `username` to list codespaces for (used with --org)")
 	cmdutil.AddJSONFlags(listCmd, &exporter, api.CodespaceFields)
 
+	listCmd.Flags().BoolVarP(&opts.useWeb, "web", "w", false, "List codespaces in the web browser, can only be used with --repo flag")
```
It might just be me, but this seems to read that `--repo` MUST be specified for `--web`. I think the error message like you used in `create` is easier to understand `cannot be used with --user or --org`

doaortu replied:
fixed in [d167a17](https://github.com/cli/cli/pull/7288/commits/d167a1761bb271b3e05919ad9edf0a237adeb7d2)

doaortu committed:
Merge branch 'cli:trunk' into feat/web-flag-for-cs-ls-create

doaortu committed:
refactor: remove the opts equality checks

doaortu committed:
fix: mutually exclusive misfires
because of wrong logic

+ refine test case too

doaortu committed:
cleanup:removing useWeb check in fn getMachineName

because it's no longer needed

+ remove redundant test-case

doaortu committed:
refactor: remove redundant ifs

doaortu committed:
refactor: clarify test name

doaortu committed:
re-clarify web flag desc in list.go

cmbrose commented on pkg/cmd/codespace/list.go:
```diff
@@ -34,11 +35,21 @@ func newListCmd(app *App) *cobra.Command {
 		Args:    noArgsConstraint,
-		RunE: func(cmd *cobra.Command, args []string) error {
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			err := cmdutil.MutuallyExclusive("using `--org` or `--user` with `--repo` is not allowed", opts.repo != "", opts.orgName != "" || opts.userName != "")
+			if err != nil {
+				return err
+			}
+			err = cmdutil.MutuallyExclusive("using `--web` with `--org` or `--user` is not supported, please use with `--repo` instead", opts.useWeb, opts.orgName != "" || opts.userName != "")
+			if err != nil {
+				return err
+			}
```
Nit: break these long lines up, use the more idiomatic `if err := ...; err != nil` technique, and use the same error format for each:

```suggestion
            if err := cmdutil.MutuallyExclusive(
                "using `--org` or `--user` with `--repo` is not allowed", 
                opts.repo != "",
                opts.orgName != "" || opts.userName != "",
            ); err != nil {
                return err
            }
            
            if err := cmdutil.MutuallyExclusive(
                "using `--org` or `--user` with `--web` is not allowed, only `--repo` is allowed", 
                opts.useWeb, 
                opts.orgName != "" || opts.userName != "",
            ); err != nil {
                return err
            }
```

doaortu replied:
Understood.
fixed in [ce9ce98](https://github.com/cli/cli/pull/7288/commits/ce9ce984ba443f6421fdd077b3ba3e72cd8fff57)

cmbrose commented on pkg/cmd/codespace/create.go:
```diff
@@ -78,11 +79,16 @@ func newCreateCmd(app *App) *cobra.Command {
 		Short: "Create a codespace",
 		Args:  noArgsConstraint,
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			return cmdutil.MutuallyExclusive("using --web with --display-name, --idle-timeout, or --retention-period is not supported", opts.useWeb, opts.displayName != "" || opts.idleTimeout != 0 || opts.retentionPeriod.Duration != nil)
```
Nit: break up this long line

```suggestion
            return cmdutil.MutuallyExclusive(
                "using --web with --display-name, --idle-timeout, or --retention-period is not supported",
                opts.useWeb, 
                opts.displayName != "" || opts.idleTimeout != 0 || opts.retentionPeriod.Duration != nil,
            )
```

cmbrose commented on pkg/cmd/codespace/create_test.go:
```diff
@@ -382,17 +426,114 @@ Alternatively, you can run "create" with the "--default-permissions" option to c
+			},
+			wantStderr: "  ✓ Codespaces usage for this repository is paid for by monalisa\n",
+			wantURL:    fmt.Sprintf("https://github.com/codespaces/new?repo=%d&ref=%s&machine=%s&location=%s", 123, "custom", "GIGA", "EastUS"),
+		},
```
Can we have a test for a wrong machine type?

doaortu replied:
Okay,
added in [f55fd79](https://github.com/cli/cli/pull/7288/commits/f55fd79142632e36cbd28e4f3add2fd2c3f898a7)

cmbrose commented on pkg/cmd/codespace/list_test.go:
```diff
@@ -1,15 +1,68 @@
+		{
+			name:     "list codespaces, negative --limit flag",
+			args:     "--limit -1",
+			wantsErr: fmt.Errorf("invalid limit: -1"),
```
Thanks for adding this missing test coverage!

doaortu replied:
You're welcome.
Thank you for detailed review

cmbrose reviewed (commented):
This is really coming together 🚀 Just a couple last style nits and we can merge this!

doaortu committed:
refactor: break long lines, use more idiomatic err

doaortu committed:
add test case for nonexistent/wrong machine

cmbrose reviewed (approved):
Looks good! Thanks for working through the iterations!

doaortu commented:
Thank you again for the detailed review, and for bearing with me until the end.
